Orange/widgets/data/owimpute.py

```diff
@@ -220,192 +104,163 @@ def __init__(self):
         method_layout.addStretch(2)
+
+        reset_button = QtGui.QPushButton(
+                "Restore all to default", checked=False, checkable=False,
+                clicked=self.reset_variable_methods, default=False,
+                autoDefault=False)
         method_layout.addWidget(reset_button)
-        self.varmethodbox = methodbox
-        self.varbgroup = bgroup
+
+        self.variable_button_group = button_group
 
         box = gui.auto_commit(
             self.controlArea, self, "autocommit", "Commit",
             orientation="horizontal", checkbox_label="Commit on any change")
         box.layout().insertSpacing(0, 80)
         box.layout().insertWidget(0, self.report_button)
+
         self.data = None
-        self.learner = None
+        self.modified = False
+        self.default_method = self.METHODS[self.default_method_index]
+        self.set_learner(None)
+
+    @property
+    def default_method_index(self):
+        return self._default_method_index
+
+    @default_method_index.setter
+    def default_method_index(self, index):
+        if self._default_method_index != index:
+            self._default_method_index = index
+            self.default_button_group.button(index).setChecked(True)
+            self.default_method = self.METHODS[self.default_method_index]
+            self.METHODS[self.DEFAULT].method = self.default_method
+
+            # update variable view
+            for index in map(self.varmodel.index, range(len(self.varmodel))):
+                self.varmodel.setData(index,
+                                      self.variable_methods.get(index.row(), self.METHODS[self.DEFAULT]),
+                                      Qt.UserRole)
+            self._invalidate()
 
     def set_default_method(self, index):
+        """Set the current selected default imputation method.
         """
-        Set the current selected default imputation method.
-        """
-        if self.default_method != index:
-            self.default_method = index
-            self.defbggroup.button(index).setChecked(True)
-            self._invalidate()
+        self.default_method_index = index
 
     @check_sql_input
     def set_data(self, data):
         self.closeContext()
-        self.clear()
+        self.varmodel[:] = []
+        self.variable_methods = {}
+        self.modified = False
         self.data = data
+
         if data is not None:
             self.varmodel[:] = data.domain.variables
             self.openContext(data.domain)
-            self.restore_state(self.variable_methods)
             itemmodels.select_row(self.varview, 0)
+
         self.unconditional_commit()
 
     def set_learner(self, learner):
         self.learner = learner
+        self.METHODS[self.MODEL_BASED_IMPUTER].learner = learner
+
+        button = self.default_button_group.button(self.MODEL_BASED_IMPUTER)
+        variable_button = self.variable_button_group.button(self.MODEL_BASED_IMPUTER)
+        if self.learner is None:
```

Comment:
I would prefer a default learner. This way, everything "just works", and if user wants to change the learner, he can.